hadoop-hdds/docs/content/design/full-volume-handling.md

```diff
@@ -0,0 +1,65 @@
+---
+title: Full Volume Handling
+summary: Immediately trigger Datanode heartbeat on detecting full volume
+date: 2025-05-12
+jira: HDDS-12929
+status: Design 
+author: Siddhant Sangwan, Sumit Agrawal
+---
+
+<!--
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License. See accompanying LICENSE file.
+-->
+
+## Summary
+On detecting a full Datanode volume during write, immediately trigger a heartbeat containing the latest storage report.
+
+## Problem
+When a Datanode volume is close to full, the SCM may not be immediately aware because storage reports are only sent 
+to it every thirty seconds. This can lead to the SCM allocating multiple blocks to containers on a full DN volume, 
+causing performance issues when the write fails. The proposal will partly solve this problem.
```

Comment:
SCM will check the container size before allocating block for this container.  Currently the container size is reported when container report is full, or container state is changed from open to other state. So SCM is kindly allocating the blocks blindly.  In this 30s storage reports, I think we should consider report the open containers too, to help SCM better understand the open container state and avoid over allocate blocks for one container.   @siddhantsangwan , what do you think?

Reply:
So you mean include a full container report for all containers in the DN, not just the ones on the full volume? We can use the method `StateContext#getFullContainerReportDiscardPendingICR`.

Reply:
Open containers report, not full container report. As only open container will grow its size. Other container's size will just shrink if there is any change.